rust/main/hyperlane-base/src/server/merkle_tree_insertions.rs

```diff
@@ -0,0 +1,51 @@
+use axum::http::StatusCode;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    db::{HyperlaneDb, HyperlaneRocksDB},
+    server::utils::{ServerErrorBody, ServerErrorResponse},
+};
+
+use super::utils::ServerResult;
+
+/// Merkle tree insertion
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct TreeInsertion {
+    /// index of the merkle insertion
+    pub leaf_index: u32,
+    /// id of the message
+    pub message_id: String,
+}
+
+/// fetch merkle tree insertions from rocksdb
+/// inclusive
+pub async fn fetch_merkle_tree_insertions(
+    db: &HyperlaneRocksDB,
+    leaf_index_start: u32,
+    leaf_index_end: u32,
+) -> ServerResult<Vec<TreeInsertion>> {
+    let mut merkle_tree_insertions =
+        Vec::with_capacity((leaf_index_end + 1 - leaf_index_start) as usize);
```

Comment:
nit: i know we do some input validation on this anyways prior to calling it, but just to be safe wonder if it's worth a saturating_sub? or maybe putting the input validation inside this function instead